### DISCUSSION TITLE
BelongsToMany: adds hyphen between 2 word model like this App\Course::long-features()

### DISCUSSION THREAD

**Discussion by rbijkercom** [laravel/nova-issues#3324] [Q&A]
- Laravel Version: 8.35.1
- Nova Version: 3.22.0
- PHP Version: 7.4
- Database Driver & Version: mysql
- Operating System and Version: Mac OS 11.2.3
- Browser type and version: Chrome 89.0.4389.114
- Reproduction Repository: https://github.com/###/###

### Description:
I have a model "Course" with a relationship `BelongsToMany::make('LongFeatures')`
When attaching a new LongFeature Nova gives an error
`Call to undefined method App\Course::long-features()`
but is should be looking for `App\Course::longFeatures()` instead.
How do I get Nova to attach this longFeature correctly?

### Detailed steps to reproduce the issue on a fresh Nova installation:
Model Course.php has this relationship
```
public function longFeatures()
{
  return $this->belongsToMany( LongFeature::class, 'long_feature_course', 'course_id')
    ->using( LongFeatureCourse::class)->orderBy( 'sort_order' );
}
```
As a Nova resource I have Nova/Course.php with this relationship:
`BelongsToMany::make('LongFeatures')`

In Nova, on the Course resource view, I can see the related longFeatures correctly.
<img width="990" alt="Screenshot 2021-04-03 at 12 06 41" src="https://user-images.githubusercontent.com/4490603/113487934-2249f700-94bb-11eb-9363-7be41d12b3df.png">

But when trying to attach a new longFeature I get this error:
`Call to undefined method App\Course::long-features()`

I get the error on the page where it is trying to attach the model:
`/nova/resources/courses/1/attach/long-features?viaRelationship=longfeatures&polymorphic=0`
<img width="1235" alt="Screenshot 2021-04-03 at 12 04 39" src="https://user-images.githubusercontent.com/4490603/113487933-20803380-94bb-11eb-9d1c-c16396e7d9db.png">

So it is putting a hyphen between the 2 words. It's looking for `App\Course::long-features()`, but it should look `App\Course::longFeatures()`

I've already tried this without luck:
`BelongsToMany::make('LongFeatures', 'longFeatures', 'App\Nova\LongFeature'),`

**Comment by crynobone** [comment]
Using below should solve your issue:

```php
BelongsToMany::make('LongFeatures', 'longFeatures')
```

**Reply by rbijkercom** [reply]
I've tried this but `BelongsToMany::make('LongFeatures', 'longFeatures'),` but it doesn't seem to work. This is my relationship setup:

App/Course.php
```
  public function longFeatures()
  {
    return $this->belongsToMany( LongFeature::class, 'long_feature_course', 'course_id')
      ->using( LongFeatureCourse::class)
      ->orderBy( 'sort_order' );
  }
```
App/LongFeature.php
```
  public function courses()
  {
    return $this->belongsToMany( Course::class, 'long_feature_course', 'long_feature_id' )
      ->using( LongFeatureCourse::class)
      ->orderBy( 'sort_order' );
  }
```
App/LongFeatureCourse.php
```
class LongFeatureCourse extends Pivot implements Sortable
{
.......
}
```
Nova/Course.php
```
BelongsToMany::make('LongFeatures', 'longFeatures'),
```
Nova/LongFeature.php
```
BelongsToMany::make('Courses'),
```

When I go to the Nova LongFeature resource page I can attach a course. No problem.

When I'm on the Course resource page I can see the related LongFeatures but when I try to attach a new one from this page I get the error.
`Call to undefined method App\Course::long-features()`

Did I do something wrong in my relationship setup?

**Reply by crynobone** [reply]
You should check if you have any 3rd party packages that might change Nova default behavior.

**Comment by crynobone** [comment]
![image](https://user-images.githubusercontent.com/172966/113556720-74ac2480-962f-11eb-9045-b580c12a38f0.png)
![image](https://user-images.githubusercontent.com/172966/113556737-7bd33280-962f-11eb-9f76-8aed22838284.png)

Tested again by reproducing above scenario on fresh Laravel Nova and unable to replicate. See https://github.com/nova-issues/issue-3324/commit/d87a9140ef094fb96841bfb7890e81f9a038c03c

**Reply by rbijkercom** [reply]
Thanks for trying to reproduce. 
The issue comes when the model has multiple words. Like LongFeature.
In your case the models only have 1 word. User and Role. 
Could you try it with a model like: FilmRole ?

**Reply by crynobone** [reply]
> Could you try it with a model like: FilmRole ?

That's totally irrelevant to what you experiencing. A relationship name doesn't depend on Eloquent model name.

**Reply by rbijkercom** [reply]
But somehow this seems to be the problem. 
I have no issues with the relationship `Course` and `Benefit`.

But it is an issue with any 2 word models:
`Course` and `ForBenefit`
`Course` and `LongFeature`

They are all set up in the same way. Still, 2 words give the error message:
Call to undefined method App\Course::long-features()
Call to undefined method App\Course::for-benefits()

**Reply by crynobone** [reply]
![image](https://user-images.githubusercontent.com/172966/113568093-6f0c0a00-9642-11eb-94f5-300e12065eba.png)

Tested and as I expected, no issue. Instead of trying to guess where the bug with on Laravel Nova. I recommend you to look at your application to see if there any customization or 3rd party packages that could cause this as it not reproducible on a fresh installation.

See https://github.com/nova-issues/issue-3324/commit/1381eb6719e204388c6fd07959f681f906ce1031

**Reply by rbijkercom** [reply]
OK, I think you are right. I can't reproduce on a clean install either. Will investigate further. 🙏
